core/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -63,12 +83,32 @@ public RestRequest(NamedXContentRegistry xContentRegistry, String uri) {
             RestUtils.decodeQueryString(uri, pathEndPos + 1, params);
         }
         this.params = params;
+        this.headers = Collections.unmodifiableMap(headers);
+        final List<String> contentType = getAllHeaderValues("Content-Type");
+        final XContentType xContentType = parseContentType(contentType);
+        if (xContentType != null) {
+            this.xContentType.set(xContentType);
+        }
     }
 
-    public RestRequest(NamedXContentRegistry xContentRegistry, Map<String, String> params, String path) {
+    /**
+     * Creates a new RestRequest
+     * @param xContentRegistry the xContentRegistry to use when parsing XContent
+     * @param params the parameters of the request
+     * @param path the path of the request. This should not contain request parameters
+     * @param headers a map of the headers. This map should implement a Case-Insensitive hashing for keys as HTTP header names are case
+     *                insensitive
+     */
+    public RestRequest(NamedXContentRegistry xContentRegistry, Map<String, String> params, String path, Map<String, List<String>> headers) {
         this.xContentRegistry = xContentRegistry;
         this.params = params;
         this.rawPath = path;
+        this.headers = Collections.unmodifiableMap(headers);
+        final List<String> contentType = getAllHeaderValues("Content-Type");
+        final XContentType xContentType = parseContentType(contentType);
+        if (xContentType != null) {
```

Comment:
This is needed because the xContentType is actually a `SetOnce` so we do not want to set it to null here since we have a backwards compatibility layer for auto-detection, which will set the auto-detected xcontent type.